controllers/clusterversion_controller.go

```diff
@@ -436,3 +464,82 @@ func (c *ClusterVersionReconciler) ensureConsolePlugin() error {
 
 	return nil
 }
+
+func (c *ClusterVersionReconciler) registerWebhook() error {
+	whConfig := &admrv1.ValidatingWebhookConfiguration{}
+	whConfig.Name = "subscription.ocs.openshift.io"
+
+	err := c.createOrUpdate(whConfig, func() error {
+
+		// openshift fills in the ca on finding this annotation
+		whConfig.Annotations = map[string]string{
+			"service.beta.openshift.io/inject-cabundle": "true",
+		}
+
+		// take a copy of the template
+		wh := *templates.SubscriptionValidatingWebhook.DeepCopy()
+		if len(whConfig.Webhooks) != 0 {
+			// do not mutate CA bundle that was injected by openshift
+			wh.ClientConfig.CABundle = whConfig.Webhooks[0].ClientConfig.CABundle
+		}
+
+		// only send requests received from own namespace
+		wh.NamespaceSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"kubernetes.io/metadata.name": c.OperatorNamespace,
+			},
+		}
+
+		// only send resources matching the label
+		wh.ObjectSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				subscriptionLabelKey: subscriptionLabelValue,
+			},
+		}
+
+		// send request to the service running in own namespace
+		wh.ClientConfig.Service.Namespace = c.OperatorNamespace
+
+		// set the webhook
+		whConfig.Webhooks = []admrv1.ValidatingWebhook{wh}
```

Comment:
The way that this function is written creates unnecessary copies of deeply nested structures in memory. This can be avoided:

```suggestion
		var CABundle []byte
		if len(whConfig.Webhooks) == 0 {
			whConfig.Webhooks = make([]admrv1.ValidatingWebhook, 1)
		} else {
			CABundle := wh.ClientConfig.CABundle
		} 

		var wh *admrv1.ValidatingWebhook = whConfig.Webhooks[0]
		templates.SubscriptionValidatingWebhook.DeepCopyInto(wh)

		wh.Name = whConfig.Name

		// only send requests received from own namespace
		wh.NamespaceSelector = &metav1.LabelSelector{
			MatchLabels: map[string]string{
				"kubernetes.io/metadata.name": c.OperatorNamespace,
			},
		}

		// only send resources matching the label
		wh.ObjectSelector = &metav1.LabelSelector{
			MatchLabels: map[string]string{
				subscriptionLabelKey: subscriptionLabelValue,
			},
		}

		// Make sure we preserve the existing CA bundle if any
		wh.ClientConfig.CABundle = CABundle

		// send request to the service running in own namespace
		wh.ClientConfig.Service.Namespace = c.OperatorNamespace
```

Reply:
done.